src/test/test-udev-rules-util.c

```diff
@@ -0,0 +1,180 @@
+#include <stdlib.h>
+#include <string.h>
+
+#include "macro.h"
+#include "string-util.h"
+#include "udev-util.h"
+
+
+static void test_parse_value(void) {
+        /* input: "valid operand"
+         * parsed: valid operand
+         * use the following command to help generate textual C strings:
+         * python3 -c 'import json; print(json.dumps(input()))' */
+        char str[] = "\"valid operand\"";
```

Comment:
Hmm, I do not like to modify the constant string... How about to introduce the following helper function.
```
static void test_udev_rule_parse_value_one(const char *in, const char *expected_value, int expected_retval) {
        _cleanup_free_ char *str = NULL;
        char *value = UINT_TO_PTR(0x12345678U);
        char *endpos = UINT_TO_PTR(0x87654321U);

        assert_se(str = strdup(in));
        assert_se(udev_rule_parse_value(...) == expected_retval);
        if (expected_retval < 0) {
                /* not modified on failure */
                assert_se(value == UINT_TO_PTR(0x12345678U));
                assert_se(endpos == UINT_TO_PTR(0x87654321U));
        } else {
                assert_se(streq_ptr(value, expected_value));
                assert_se(endpos == str + strlen(in) - 1);
                assert_se(*endpos == '\0');
        }
}
```

Reply:
`endpos` is not necessarily zero though. The position where the input string ends, e.g. the position the parsing should continue will not be modified if the output string is shorter.

Reply:
I've moved the + 1 from udev-rule from `udev-rules.c:parse_line()` into `udev_rule_parse_value`